virttest/qemu_devices/qcontainer.py

```diff
@@ -2262,3 +2314,22 @@ def input_define_by_params(self, params, name, bus=None):
             logging.warn("'%s' is not supported by your qemu", driver)
 
         return devices
+
+    def vcpu_device_define_by_params(self, params, name):
+        """
+        Create vcpu device by params.
+        :param params: vpcu device params.
+        :param name: Object name of input device.
+        """
+        params = params.object_params(name)
+        vcpu_driver = params["vcpu_driver"]
+        if not self.has_device(vcpu_driver):
+            raise virt_vm.VMDeviceNotSupportedError(self.vmname, vcpu_driver)
+        vcpu_params = {"id": params.get("vcpu_id", name)}
+        vcpu_params.update(params.copy_from_keys(params.objects("vcpu_props")))
```

Comment:
Well, that means we must set the vcpu properties by using the option names defined by qemu, e.g.
```
core-id = 1
socket-id = 2
kvm_mmu = yes
...
```
this may result in name conflicts with other params.

Suggest either:

* changing the data type of the `vcpu_props`  param to json or dict

or 

* putting each property into individual one

Reply:
Actually, this method uses `object_params` to format at first, so I think it should not conflicts with other params. But I pick your suggestion, looks like json format is more readable.